### PR TITLE
HIVE-24715 Increase bucketId range (amagyar)

### PR DESCRIPTION
**Pull request by zeroflag**
Currently the bucketId field is stored on 12 bits. When TEZ starts more tasks than 4095 it overflows.

The proposed solution is to let the bucket id overflow into the statement id. Depending on the maximum statement id (we  know that at compile time) we can dynamically allocate bits for the bucket id from the statement id field.

For example in a single statement operation only one bit is needed for the statement id and 11 extra bits can be allocated for the bucket id.

In case of a multi insert where the max statement id lets say 100 (0b000001100100), five extra bits can be allocated for the bucket Id.

Compaction ignores statement ids therefore the same buckets (delta_1_1_0/bucket_00904_0 and delta_1_1_178/bucket_00904_0) will be merged together after running a major compaction, so we won’t end up having more than 4096 files in the end.

After running an INSERT with 6000 parallel tasks we end up having 2 delta directories:

```
/warehouse/tablespace/managed/hive/cdr5bi/delta_0000039_0000039_0000
/warehouse/tablespace/managed/hive/cdr5bi/delta_0000039_0000039_0002  
```

Where in the first we have buckets from 0 .. 4095 and in the second we have buckets from 0 .. 1903.
After running compaction we expect having a single base directory with buckets from 0 .. 4095.

```
/warehouse/tablespace/managed/hive/cdr5bi/base_0000039_v0003975
```

The same buckets from different statements are merged together.

See more info in the attached pdf in the Jira.

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/io/TestBucketCodec.java [hunk 1]
```diff
@@ -77,13 +77,13 @@ public void testGetBucketCodecVersion1() {
 
   @Test(expected = IllegalArgumentException.class)
   public void testGetBucketCodecVersion1EncodeNegativeBucketId() {
-    BucketCodec.getCodec(1).encode(new AcidOutputFormat.Options(null).bucket(-1).statementId(16));
+    BucketCodec.getCodec(1).encode(new AcidOutputFormat.Options(null).bucket(-1).statementId(16).maxStmtId(BucketCodec.MAX_STATEMENT_ID));
```
Comment (pvargacl) [inline ql/src/test/org/apache/hadoop/hive/ql/io/TestBucketCodec.java:80]: Why is this needed? This is the default behaviour if maxstatementId was not set, no?

**Comment by pvargacl** [comment]
I think it could be useful to have a configuration property that could disable this behaviour. If I understand correctly it would be enough to not set maxstatementId in filesink when it is disabled, and it will fallback to the old way.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 2]
```diff
@@ -4509,7 +4509,6 @@ public static void handleDirectInsertTableFinalPath(Path specPath, String unionS
         fs.mkdirs(specPath);
       }
     }
-
```
Comment (pvargacl) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4512]: nit, remove this change

Comment (zeroflag) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java]: ok

File: ql/src/test/org/apache/hadoop/hive/ql/io/TestBucketCodec.java (hunk 1 above)
Comment (zeroflag) [reply ql/src/test/org/apache/hadoop/hive/ql/io/TestBucketCodec.java]: Right, this is no longer needed. I removed it.

**Comment by zeroflag** [comment]
@pvargacl 

> I think it could be useful to have a configuration property that could disable this behaviour. If I understand correctly it would be enough to not set maxstatementId in filesink when it is disabled, and it will fallback to the old way.

Ok, I added a config with a default true value.

**Review by pvargacl — APPROVED** [review]
LGTM +1
